modules/core/src/main/java/org/apache/ignite/internal/util/GridConcurrentMultiPairQueue.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.util;
+
+import org.apache.ignite.internal.util.typedef.T2;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Concurrent queue that wraps collection of {@code Pair<K, V[]>}
+ * The only garantee {@link #poll} provided is sequentially emtify values per key array.
+ * i.e. input like: <br>
+ * p1 = new Pair<1, [1, 3, 5, 7]> <br>
+ * p2 = new Pair<2, [2, 3]> <br>
+ * p3 = new Pair<3, [200, 100]> <br>
+ * and further sequence of {@code poll} calls may produce output like: <br>
+ * [3, 200], [3, 100], [1, 1], [1, 3], [1, 5], [1, 7], [2, 2], [2, 3]
+ *
+ * @param <K> The type of key in input pair collection.
+ * @param <V> The type of value array.
+ */
+public class GridConcurrentMultiPairQueue<K, V> {
```

Comment:
What do you think about the following idea:
Instead of class like this which contains key(PageMemory) and value(Pages) create a class which would be returned iterator over all PageMemory which in turn would be returned iterator over pages.

Something like that
```
PageMemory memory = pageMemories.poll();
while(memory != null){
   FullPageId page = memory.poll();
   while(page != null){
      ...
      page = memory.poll();
   }
   memory = pageMemories.poll();
}
```

This method, at least, helps to avoid wrapping in T2 on every page during iteration.

Reply:
Another way to approach this problem is to pass handler into the queue and it will be internally iterated, like "forEach((a, b) -> doSomeAction(a, b))".

Reply:
fixed using @ibessonov below approach.